### ISSUE TITLE
shard_map with a closed over argument produces the wrong batched axis size when using spmd_axis_name

### ISSUE THREAD

**Issue by shoyer** [google/jax#17874]
### Description

To reproduce:
```python
import chex
import jax
from jax.experimental import shard_map
import numpy as np
import jax.numpy as jnp
import functools

chex.set_n_cpu_devices(10)

shmap = shard_map.shard_map
P = jax.sharding.PartitionSpec

mesh = jax.sharding.Mesh(jax.devices(), axis_names=['x'])


@functools.partial(jax.vmap, spmd_axis_name='x')
def shmap_explicit_arg(x):
  return shmap(lambda x: x, mesh, (P(),), P())(x)

shape = shmap_explicit_arg(jnp.zeros((10,))).shape
assert shape == (10,), shape


@functools.partial(jax.vmap, spmd_axis_name='x')
def shmap_closed_arg(x):
  return shmap(lambda: x, mesh, (), P())()

shape = shmap_closed_arg(jnp.zeros((10,))).shape
assert shape == (10,), shape
```

The first assertion with the explicit argument passes.

The second assertion with the closed over argument fails, with `shape = (100,)`. If remove `spmd_axis_name` from the `vmap` call it passes.
